### PR TITLE
Create Blog Catalog

### PR DESCRIPTION
**Pull request by ppalashturov**
Create Blog Catalog for OpenCart 4

### CODE REVIEW

**Comment by TheCartpenter** [comment]
While I am not sure about the utility of this feature in the core, the array keys and values in your ```filter_data``` array are not aligned.

**Comment by ppalashturov** [comment]
> While I am not sure about the utility of this feature in the core, the array keys and values in your `filter_data` array are not aligned.

Hello, Could you give me a hint where the problem is or screenshot?
Best Regards

**Comment by danielkerr** [comment]
i will add a bloog my self. looks like u followed my code style well though

**Comment by ppalashturov** [comment]
> i will add a bloog my self. looks like u followed my code style well though

Thank you!

**Comment by stalker780** [comment]
It would be better to call blog entity as article and blog_category as topic.

This will have more sense and will not confuse anyone with product categories.

Catalog
- Categories
- Products
- Manufacturers
- Reviews

Blog
- Topics
- Articles
- Authors
- Comments

**Comment by stalker780** [comment]
Blog is a must have in modern OpenCart store. For news, product reviews.
It is very strange OC still does not have it. Though there are dozens of different blog extensions.

**Comment by TheCartpenter** [comment]
I just hope the ```customer_id``` or the ```customer_group_id``` field won't be integrated in the blog's main table but rather the ```sku``` field in the main table as to have the ```customer_id``` or ```customer_group_id``` in the ```activity``` logs since APIs will still be able to track down those information.

**Comment by ppalashturov** [comment]
@danielkerr and @stalker780 I just rewrote the blog module according to the idea of @stalker780. Please check now! I have to add the comments but that will be tomorrow.

**Comment by TheCartpenter** [comment]
Exactly what I was referring on the above. You are capturing the customer ID rather than the product SKU: https://github.com/opencart/opencart/pull/12507/commits/e76803eef281a87182dd7ca2aaaa63bfa8a1555b#diff-8e21dfc3a7ec4a95d4f6a422c54a881cfa2df331499e4b4ef45b022aa4d5fb51R3467 .

**Comment by ppalashturov** [comment]
> Exactly what I was referring on the above. You are capturing the customer ID rather than the product SKU: [e76803e#diff-8e21dfc3a7ec4a95d4f6a422c54a881cfa2df331499e4b4ef45b022aa4d5fb51R3467](https://github.com/opencart/opencart/commit/e76803eef281a87182dd7ca2aaaa63bfa8a1555b#diff-8e21dfc3a7ec4a95d4f6a422c54a881cfa2df331499e4b4ef45b022aa4d5fb51R3467) .

Sorry, but I can't understand what you mean. The blog comments are not related to the product SKU.

**Comment by TheCartpenter** [comment]
If you intend to create your blogs without any product relationships, then capturing the customer ID directly with the store independently will only encourage greater marketing levels with the use of their APIs.

**Comment by TheCartpenter** [comment]
It would be great to also include email event notifications to store owners. In your code, no mail objects are use to notify store owners at this time.

**Comment by TheCartpenter** [comment]
The latest integration would become quite complexed to manage since using ```article``` is normally used as a singular term as we would be falling back into the same situation we used to be with the ```recurring``` and the same thing applies with the ```blog```.

@ppalashturov: Since ```recurring``` has been converted to ```subscriptions``` and that ```subscriptions``` also contains the field of: ```product_id``` from the core and that the ```catalog``` is used in ```quantitate``` whereas ```article``` is rather used as a ```qualitive``` form, and that ```orders``` gathers ```products``` at the end of the day, the question would be to know how would you convert the ```article_id``` with the ```orders``` since ```orders``` are being delivered as a ```qualitive``` form as well?

As you mentioned on the above, a ```blog``` and, in this case, an ```article``` has nothing to do with ```products``` but ```orders``` do take ```products``` at the end of the day from the core since the beginning. Therefore, how would you deliver the purpose of storing the ```article_id``` with the ```product_id``` is already stored with the ```orders``` for review purposes?

**Comment by TheCartpenter** [comment]
Besides, to simplify manners from the above, the ```subscription plans``` can already be used in the ```catalog``` as for the ```options``` that can already be converted into ```product options```. Therefore, the same concept should be used with the ```articles``` since we're also using ```reviews``` in the ```catalog```.

**Comment by TheCartpenter** [comment]
The codes mail/comment are also not aligned.

File: upload/catalog/controller/blog/comment.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+		return $this->load->view('blog/comment', $data);
+	}
+
+	public function write(): void {
+		$this->load->language('blog/comment');
+
+		$json = [];
+
+		if (isset($this->request->get['article_id'])) {
+			$article_id = $this->request->get['article_id'];
```
Comment (TheCartpenter) [inline upload/catalog/controller/blog/comment.php:57]: Replace with:

```$article_id = (int)$this->request->get['article_id'];```

File: upload/admin/language/en-gb/setting/setting.php [hunk 2]
```diff
@@ -129,6 +133,8 @@
 $_['entry_captcha_page']                  = 'Captcha Page';
 $_['entry_logo']                          = 'Store Logo';
 $_['entry_image_category']                = 'Category Image Size (W x H)';
+$_['entry_image_topic']           		  = 'Topic List Image Size (W x H)';
```
Comment (TheCartpenter) [inline upload/admin/language/en-gb/setting/setting.php:136]: Equal not aligned.

File: upload/admin/language/en-gb/setting/setting.php [hunk 3]
```diff
@@ -261,6 +269,8 @@
 $_['error_image_popup']                   = 'Product Image Popup Size dimensions required!';
 $_['error_image_product']                 = 'Product List Size dimensions required!';
 $_['error_image_category']                = 'Category List Size dimensions required!';
+$_['error_image_topic']           = 'Category Blog List Size dimensions required!';
```
Comment (TheCartpenter) [inline upload/admin/language/en-gb/setting/setting.php:272]: Equal not aligned.

**Comment by TheCartpenter** [comment]
@ppalashturov: Since your codes are also using layouts, is it possible to add data in the ```statistics``` whereas the ```store_id``` can be related with? In this case, with each ```_layout``` database table that uses the ```store_id``` field? 

This way, no matter where the customers, guests, and search engines will visit the page from the store, the ```statistics``` data can be added for further reviews while using the ```admin reports``` for the ```articles``` and ```topics``` that are already associated with the ```stores``` from your codes and on the database.

File: upload/admin/controller/report/statistics.php [hunk 4]
```diff
@@ -218,4 +218,26 @@ public function review(): void {
...
+
+		if (!$this->user->hasPermission('modify', 'report/statistics')) {
+			$json['error'] = $this->language->get('error_permission');
+		}
+
+		if (!$json) {
+			$this->load->model('report/statistics');
+			$this->load->model('blog/comment');
+
+			$this->model_report_statistics->editValue('comment', $this->model_blog_comment->getTotalCommentsAwaitingApproval());
```
Comment (TheCartpenter) [inline upload/admin/controller/report/statistics.php:235]: This line would only pull the comments that are not approved yet vs. the ones that would be already approved.

Comment (ppalashturov) [reply upload/admin/controller/report/statistics.php:235]: That's it with the product review!

Comment (TheCartpenter) [reply upload/admin/controller/report/statistics.php:235]: Yes, correct.

**Comment by TheCartpenter** [comment]
```$this->model_blog``` should not be used as an independent form. It should rather be used as: ```$this->model_marketing_blog``` since ```content``` can also be used vs. ```marketing``` strategy on websites. The ```routes``` in the ```admin menu``` should be used as: ```marketing > blog``` and so forth since ```blog``` reflects in ```marketing``` ```models``` whereas a ```controller``` from the ```admin marketing folder``` could gather these ```models``` on anyhow.

Since the platform also includes```affiliates```, and that ```affiliates models``` are also included in ```marketing``` from the ```admin```, then the ```catalog```-end should also reflect on ```marketing``` for the ```blog``` and ```affiliates``` while ```marketing``` can also be used for way many more subjects. There used to be a ```commit``` about this on this ```repository``` before that explained 5 ```categories``` of ```marketing strategies```. However, by ```Googling``` this information, it seem there are way many more ```classifications``` that could be added in the future.

**Comment by lemonadvert** [comment]
Hello,
When are you planning to add the Blog functionality to OpenCart 4? 
It will be great if we have a Blog integrated into OpenCart.
There will be no need to add additional blog modules.

**Comment by pemaier** [comment]
Hello.
To have a Blog is a good idea and it is also necessary in the basis/standard Opencart version in my opinion. (even if some devs will loose business)
And if it is implemented it should not have only a minimum basic function set. 
It should also be possible to direct the articles to different stores and customer groups. (Stores is already implemented I see.) The experience showed me in the past that shops with multistore architecture have mostly also different customer groups. So the work with blogs should also serve the requirement to direct article categories to one or several customer groups. If this is also necessary for the articles ... I'm not sure.

**Comment by stalker780** [comment]
@mhcwebdesign it is a very weak argument.
Show me any good store without news/blog.

It's like: a car is for transporting ones ass from point A to point B. Who the heck may need automatic transmission, conditioner, electric window lifter and mp3 player?

**Comment by pemaier** [comment]
**Thank you @ppalashturov** for your time invest to bring this topic really further.
But I'm also more or less sure this is not the last solution what we will find the way into the master branch. 
As @danielkerr was writing he plans to make a own solution and because some additional coding is required (e,g. phpdocs) it will be a good base to add some further additions to meet OpenCart standards. ... I only hope it will be included the next version. :)

**Comment by ppalashturov** [comment]
Thanks @pemaier :) I also added phpdocs to the blog!

File: upload/admin/controller/setting/setting.php [hunk 5]
```diff
@@ -394,6 +397,11 @@ public function index(): void {
 			'value' => 'review'
 		];
```
Comment (TheCartpenter) [inline upload/admin/controller/setting/setting.php:399]: ```$data['captcha_pages'] = [];```

File: upload/admin/controller/setting/setting.php [hunk 6]
```diff
@@ -584,6 +616,11 @@ public function index(): void {
...
```
Comment (TheCartpenter) [inline upload/admin/controller/setting/setting.php:618]: ```$data['mail_alerts'] = [];```

File: upload/admin/language/en-gb/marketplace/promotion.php [hunk 7]
```diff
@@ -4,3 +4,6 @@
 $_['text_install']     = 'Install';
 $_['text_uninstall']   = 'Uninstall';
 $_['text_delete']      = 'Delete';
+
+// Column
+$_['column_name'] = 'Recommended Name';
```
Comment (TheCartpenter) [inline upload/admin/language/en-gb/marketplace/promotion.php:9]: Equal needs to be aligned.

File: upload/admin/language/en-gb/setting/setting.php (hunk 2 above)
Comment (TheCartpenter) [inline upload/admin/language/en-gb/setting/setting.php:136]: Equal needs to be aligned.

File: upload/admin/model/blog/article.php [hunk 8]
```diff
@@ -0,0 +1,307 @@
+<?php
+
+namespace Opencart\Admin\Model\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:5]: Needs a description.

File: upload/admin/model/blog/article.php [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+		$this->db->query("DELETE FROM `" . DB_PREFIX . "article_to_layout` WHERE `article_id` = '" . (int)$article_id . "'");
+		$this->db->query("DELETE FROM `" . DB_PREFIX . "seo_url` WHERE `key` = 'article_id' AND `value` = '" . (int)$article_id . "'");
+
+		$this->db->query("DELETE FROM `" . DB_PREFIX . "article_comment` WHERE `article_id` = '" . (int)$article_id . "'");
+
+		$this->cache->delete('article');
+	}
+
+	/**
+	 * @param int   $article_id
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:118]: Same here.

File: upload/admin/model/blog/article.php [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+	 * @param int   $rating
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:119]: Same here.

File: upload/admin/model/blog/article.php [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+	 *
+	 * @return void
+	 */
+	public function editRating(int $article_id, int $rating): void {
+		$this->db->query("UPDATE `" . DB_PREFIX . "article` SET `rating` = '" . (int)$rating . "', `date_modified` = NOW() WHERE `article_id` = '" . (int)$article_id . "'");
+	}
+
+	/**
+	 * @param int   $article_id
+	 *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:129]: Same here.

File: upload/admin/model/blog/article.php [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+	 */
+	public function getArticle(int $article_id): array {
+		$query = $this->db->query("SELECT DISTINCT * FROM `" . DB_PREFIX . "article` WHERE `article_id` = '" . (int)$article_id . "'");
+
+		return $query->row;
+	}
+
+	/**
+	 * @param array   $data
+	 *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:140]: Same here.

File: upload/admin/model/blog/article.php [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+
+			$this->cache->set('article.' . md5($sql), $article_data);
+		}
+
+		return $article_data;
+	}
+
+	/**
+	 * @param int   $article_id
+	 *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:194]: Same here.

File: upload/admin/model/blog/article.php [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+		foreach ($query->rows as $result) {
+			$article_seo_url_data[$result['store_id']][$result['language_id']] = $result['keyword'];
+		}
+
+		return $article_seo_url_data;
+	}
+
+	/**
+	 * @param int   $article_id
+	 *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/article.php:251]: Also for the rest.

File: upload/admin/model/blog/comment.php [hunk 15]
```diff
@@ -0,0 +1,196 @@
+<?php
+namespace Opencart\Admin\Model\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/comment.php:4]: Empty description.

File: upload/catalog/controller/blog/topic.php [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+				'start' => ($page - 1) * $limit,
+				'limit' => $limit
+			];
+
+			$article_total = $this->model_blog_article->getTotalArticles($filter_data);
+
+			$results = $this->model_blog_article->getArticles($filter_data);
+
+			foreach ($results as $result) {
+
```
Comment (TheCartpenter) [inline upload/catalog/controller/blog/topic.php:82]: Empty space here.

File: upload/catalog/controller/blog/topic.php [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+			$data['pagination'] = $this->load->controller('common/pagination', [
+				'total' => $article_total,
+				'page' => $page,
+				'limit' => $limit,
+				'url' => $this->url->link('blog/topic', 'language=' . $this->config->get('config_language') . '&topic_id=' . $this->request->get['topic_id'] . $url . '&page={page}')
+			]);
+
+			$data['results'] = sprintf($this->language->get('text_pagination'), ($article_total) ? (($page - 1) * $limit) + 1 : 0, ((($page - 1) * $limit) > ($article_total - $limit)) ? $article_total : ((($page - 1) * $limit) + $limit), $article_total, ceil($article_total / $limit));
+
+			// http://googlewebmastercentral.blogspot.com/2011/09/pagination-with-relnext-and-relprev.html
```
Comment (TheCartpenter) [inline upload/catalog/controller/blog/topic.php:106]: Why? The pagination from the core cannot accomplish this?

File: upload/catalog/model/blog/article.php [hunk 18]
```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Opencart\Catalog\Model\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/catalog/model/blog/article.php:5]: No description.

File: upload/catalog/model/blog/topic.php [hunk 19]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Opencart\Catalog\Model\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/catalog/model/blog/topic.php:5]: No description.

File: upload/catalog/view/template/blog/comment.twig [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+            $('#form-comment').find('.invalid-feedback').removeClass('d-block');
+
+            if (json['error']) {
+                if (json['error']['warning']) {
+                    $('#alert').prepend('<div class="alert alert-danger alert-dismissible"><i class="fa-solid fa-circle-exclamation"></i> ' + json['error']['warning'] + ' <button type="button" class="btn-close" data-bs-dismiss="alert"></button></div>');
+                }
+
+                for (key in json['error']) {
+                    $('#input-' + key.replaceAll('_', '-')).addClass('is-invalid').find('.form-control, .form-select, .form-check-input, .form-check-label').addClass('is-invalid');
+                    $('#error-' + key.replaceAll('_', '-')).html(json['error'][key]).addClass('d-block');
```
Comment (TheCartpenter) [inline upload/catalog/view/template/blog/comment.twig:83]: ```json['error']['key']```

File: upload/catalog/view/template/blog/topic.twig [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+      <h1>{{ heading_title }}</h1>
+      {{ description }}
+      {% if articles %}
+      <div class="row mb-2">
+        {% for article in articles %}
+        <div class="col-md-6">
+          <div class="row g-0 border rounded overflow-hidden flex-md-row mb-4 shadow-sm h-md-250 position-relative">
+            <div class="col p-4 d-flex flex-column position-static">
+              <h3 class="mb-0">{{ article.title }}</h3>
+{#              <div class="mb-1 text-body-secondary">Nov 12</div>#}
```
Comment (TheCartpenter) [inline upload/catalog/view/template/blog/topic.twig:20]: Trailing code.

**Comment by pemaier** [comment]
I was installing the OpenCart from opencartbulgaria:master with blog and tested it.
Here are some comments:
- the position in the menu is perfect
- the forms topic / articles / comments are basically OK
- in the articles list I'm missing the date of creation and a filter 
  // because if there are more than 30 articles it starts to not so easy to find and organize articles
- also as I said before in the topics form I'm missing the entries of customer groups
- also as I said before in the articles I'm missing a start date to view the articles

then I guess that I didn't have a complete install. I'm missing a possibility for some settings like e.g. accept comments or not.

I don't found a possibility to place the blog in modules to show it e.g. on main page (or somewhere else) 
also I don't found a link to make comments after I activated the Blog Topic to show at bottom.
At this point I stop the testing now.

File: upload/admin/language/en-gb/blog/article.php [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+$_['entry_sort_order']        = 'Sort Order';
+$_['entry_layout']            = 'Layout Override';
+$_['entry_topic']     		  = 'Topic';
+
+// Help
+$_['help_bottom']             = 'Display in the bottom footer.';
+
+// Error
+$_['error_warning']           = 'Warning: Please check the form carefully for errors!';
+$_['error_topic']     = 'Topic required!';
```
Comment (TheCartpenter) [inline upload/admin/language/en-gb/blog/article.php:40]: Equal not aligned.

File: upload/admin/language/en-gb/setting/setting.php (hunk 2 above)
Comment (TheCartpenter) [reply upload/admin/language/en-gb/setting/setting.php:136]: Still happening.

File: upload/admin/controller/blog/article.php [hunk 23]
```diff
@@ -0,0 +1,473 @@
...
+		$this->load->model('blog/topic');
+
+		$data['topics'] = [];
+
+		$results = $this->model_blog_topic->getTopics();
+
+		foreach ($results as $result) {
+			$data['topics'][] = [
+				'topic_id' => $result['topic_id'],
+				'title' => $result['title'],
```
Comment (TheCartpenter) [inline upload/admin/controller/blog/article.php:287]: Array key and value not aligned.

File: upload/admin/controller/blog/comment.php [hunk 24]
```diff
@@ -0,0 +1,581 @@
+<?php
+
+namespace Opencart\Admin\Controller\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/admin/controller/blog/comment.php:5]: No description.

File: upload/admin/controller/blog/comment.php [hunk 25]
```diff
@@ -0,0 +1,581 @@
...
+		$this->load->model('blog/comment');
+
+		$comment_total = $this->model_blog_comment->getTotalComments($filter_data);
+
+		$results = $this->model_blog_comment->getComments($filter_data);
+
+		foreach ($results as $result) {
+			$data['comments'][] = [
+				'article_comment_id'  => $result['article_comment_id'],
+				'title'       => $result['title'],
```
Comment (TheCartpenter) [inline upload/admin/controller/blog/comment.php:226]: Array keys and values not aligned.

File: upload/admin/controller/blog/topic.php [hunk 26]
```diff
@@ -0,0 +1,413 @@
+<?php
+namespace Opencart\Admin\Controller\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/admin/controller/blog/topic.php:4]: No description.

File: upload/admin/controller/blog/topic.php [hunk 27]
```diff
@@ -0,0 +1,413 @@
...
+		$topic_total = $this->model_blog_topic->getTotalTopics();
+
+		$results = $this->model_blog_topic->getTopics($filter_data);
+
+		foreach ($results as $result) {
+
+			$blog_total = $this->model_blog_article->getTotalArticlesByTopic($result['topic_id']);
+
+			$data['topics'][] = [
+				'topic_id' => $result['topic_id'],
```
Comment (TheCartpenter) [inline upload/admin/controller/blog/topic.php:123]: Array keys and values are not aligned.

File: upload/admin/controller/common/column_left.php [hunk 28]
```diff
@@ -135,6 +135,44 @@ public function index(): string {
 				];
 			}
 
+			//Blog
```
Comment (TheCartpenter) [inline upload/admin/controller/common/column_left.php:138]: ```// Blog```

File: upload/admin/language/en-gb/marketplace/promotion.php (hunk 7 above)
Comment (TheCartpenter) [reply upload/admin/language/en-gb/marketplace/promotion.php:9]: Still happening.

File: upload/admin/model/blog/article.php (hunk 8 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:5]: Still happening.

File: upload/admin/model/blog/article.php [hunk 29]
```diff
@@ -0,0 +1,307 @@
+<?php
+
+namespace Opencart\Admin\Model\Blog;
+/**
+ *
+ */
+class Article extends \Opencart\System\Engine\Model {
+	/**
+	 * @param array $data
+	 *
```
[the remark this answers is not in the document]
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:10]: Still happening.

File: upload/admin/model/blog/article.php [hunk 30]
```diff
@@ -0,0 +1,307 @@
...
+				$this->db->query("INSERT INTO `" . DB_PREFIX . "article_to_layout` SET `article_id` = '" . (int)$article_id . "', `store_id` = '" . (int)$store_id . "', `layout_id` = '" . (int)$layout_id . "'");
+			}
+		}
+
+		$this->cache->delete('article');
+	}
+
+	/**
+	 * @param int   $article_id
+	 *
```
[the remark this answers is not in the document]
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:102]: Still happening.

File: upload/admin/model/blog/article.php (hunk 9 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:118]: Still happening.

File: upload/admin/model/blog/article.php (hunk 11 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:129]: Still happening.

File: upload/admin/model/blog/article.php (hunk 12 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:140]: Still happening.

File: upload/admin/model/blog/article.php (hunk 13 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:194]: Still happening.

File: upload/admin/model/blog/article.php [hunk 31]
```diff
@@ -0,0 +1,307 @@
...
+				'meta_keyword' => $result['meta_keyword']
+			];
+		}
+
+		return $article_description_data;
+	}
+
+	/**
+	 * @param int   $article_id
+	 *
```
[the remark this answers is not in the document]
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:217]: Still happening.

File: upload/admin/model/blog/article.php [hunk 32]
```diff
@@ -0,0 +1,307 @@
...
+		foreach ($query->rows as $result) {
+			$article_store_data[] = $result['store_id'];
+		}
+
+		return $article_store_data;
+	}
+
+	/**
+	 * @param int   $article_id
+	 *
```
[the remark this answers is not in the document]
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:234]: Still happening.

File: upload/admin/model/blog/article.php (hunk 14 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/article.php:251]: Still happening.

File: upload/admin/model/blog/comment.php (hunk 15 above)
Comment (TheCartpenter) [reply upload/admin/model/blog/comment.php:4]: Still happening.

File: upload/admin/model/blog/topic.php [hunk 33]
```diff
@@ -0,0 +1,268 @@
+<?php
+namespace Opencart\Admin\Model\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/admin/model/blog/topic.php:4]: Needs a description.

File: upload/admin/view/template/blog/article_list.twig [hunk 34]
```diff
@@ -0,0 +1,38 @@
+<form id="form-article" method="post" data-oc-toggle="ajax" data-oc-load="{{ action }}" data-oc-target="#article">
+	<div class="table-responsive">
```
Comment (TheCartpenter) [inline upload/admin/view/template/blog/article_list.twig:2]: Not using indent of 2.

File: upload/admin/view/template/blog/comment_list.twig [hunk 35]
```diff
@@ -0,0 +1,40 @@
+<form id="form-comment" method="post" data-oc-toggle="ajax" data-oc-load="{{ action }}" data-oc-target="#comment">
+	<div class="table-responsive">
```
Comment (TheCartpenter) [inline upload/admin/view/template/blog/comment_list.twig:2]: Not using indent of 2.

File: upload/admin/view/template/blog/topic.twig [hunk 36]
```diff
@@ -0,0 +1,30 @@
+{{ header }}{{ column_left }}
+<div id="content">
+	<div class="page-header">
```
Comment (TheCartpenter) [inline upload/admin/view/template/blog/topic.twig:3]: Not using indent of 2.

File: upload/admin/view/template/blog/topic_list.twig [hunk 37]
```diff
@@ -0,0 +1,36 @@
+<form id="form-topic" method="post" data-oc-toggle="ajax" data-oc-load="{{ action }}" data-oc-target="#topic">
+	<div class="table-responsive">
```
Comment (TheCartpenter) [inline upload/admin/view/template/blog/topic_list.twig:2]: Not using indent of 2.

File: upload/catalog/controller/blog/article.php [hunk 38]
```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace Opencart\Catalog\Controller\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/catalog/controller/blog/article.php:5]: Needs a description.

File: upload/catalog/controller/blog/comment.php [hunk 39]
```diff
@@ -0,0 +1,181 @@
+<?php
+
+namespace Opencart\Catalog\Controller\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/catalog/controller/blog/comment.php:5]: Needs a description.

File: upload/catalog/controller/blog/topic.php [hunk 40]
```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace Opencart\Catalog\Controller\Blog;
+/**
+ *
```
Comment (TheCartpenter) [inline upload/catalog/controller/blog/topic.php:5]: Needs a description.

File: upload/catalog/controller/blog/topic.php (hunk 16 above)
Comment (TheCartpenter) [reply upload/catalog/controller/blog/topic.php:82]: Still happening.

File: upload/catalog/controller/blog/topic.php (hunk 17 above)
Comment (TheCartpenter) [reply upload/catalog/controller/blog/topic.php:106]: Unexplained. Resolved without resolution.

File: upload/catalog/model/blog/article.php (hunk 18 above)
Comment (TheCartpenter) [reply upload/catalog/model/blog/article.php:5]: Still happening.

File: upload/catalog/model/blog/comment.php [hunk 41]
```diff
@@ -0,0 +1,50 @@
+<?php
+namespace Opencart\Catalog\Model\Blog;
+/**
+ *
```
[the remark this answers is not in the document]
Comment (TheCartpenter) [reply upload/catalog/model/blog/comment.php:4]: Still happening.

File: upload/catalog/model/blog/topic.php (hunk 19 above)
Comment (TheCartpenter) [reply upload/catalog/model/blog/topic.php:5]: Still happening.

File: upload/catalog/view/template/blog/topic.twig (hunk 21 above)
Comment (TheCartpenter) [reply upload/catalog/view/template/blog/topic.twig:20]: Still there.

**Comment by danielkerr** [comment]
http://localhost/opencart-master/upload/index.php?route=cms/blog&language=en-gb&topic_id=1

**Comment by BuslikDrev** [comment]
Blog Search?

**Comment by danielkerr** [comment]
its working now

**Comment by danielkerr** [comment]
still need to finish comments

**Comment by danielkerr** [comment]
It's not finished

**Comment by ppalashturov** [comment]
@danielkerr ok :)

**Pull request closed without merging**